iothub/device/src/ClientPropertyCollection.cs

```diff
@@ -180,35 +181,72 @@ public virtual bool TryGetValue<T>(string componentName, string propertyName, ou
                     $"TryGetValue will attempt to get the property value but may not behave as expected.", nameof(TryGetValue));
             }
 
+            // If either the component name or the property name is null, empty or whitespace,
+            // then return unsuccessfully (false) with the default value of the type <T> passed in.
+            if (string.IsNullOrWhiteSpace(componentName) || string.IsNullOrWhiteSpace(propertyName))
+            {
+                propertyValue = default;
+                return false;
+            }
+
             if (Contains(componentName, propertyName))
             {
                 object componentProperties = Collection[componentName];
 
+                // If the ClientPropertyCollection was constructed by the user application (eg. for updating the client properties)
+                // then the componentProperties are retrieved as a dictionary.
+                // The required property value can be fetched from the dictionary directly.
                 if (componentProperties is IDictionary<string, object> nestedDictionary)
                 {
-                    if (nestedDictionary.TryGetValue(propertyName, out object dictionaryElement))
+                    // First verify that the retrieved dictionary contains the component identifier { "__t": "c" }.
+                    // If not, then the retrieved nested dictionary is actually a root-level property of type map.
+                    if (nestedDictionary.TryGetValue(ConventionBasedConstants.ComponentIdentifierKey, out object componentIdentifierValue)
+                        && componentIdentifierValue.Equals(ConventionBasedConstants.ComponentIdentifierValue))
```

Comment:
This seems like a string comparison, `componentIdentifierValue == ConventionBasedConstants.ComponentIdentifierValue` is preferred. And even if you choose to use .Equals, switch the order to avoid nullref exception if `componentIdentifierValue` is `null`. `ConventionBasedConstants.ComponentIdentifierValue.Equals(componentIdentifierValue)`

Reply:
Ah, I always keep on forgetting that string `==` works like value comparison in c#. I'll change this to `==`, sure.